### ISSUE TITLE
Call super constructor

### ISSUE THREAD

**Issue by ComFreek** [munificent/wren#250]
How can I call a super constructor?

```
class A {
  new {
    IO.print("A's constructor")
  }
}

class B is A {
  new {
    IO.print("B's constructor")
  }
}

new B
```
The above code will only output *B's constructor'. I've already tried adding `super.new` in B's constructor to no avail. It seems that this feature is not implemented yet.

**Comment by krypt-n** [comment]
Try using just `super`. Works for me in your example.

**Comment by gsmaverick** [comment]
You call `super()` in this case and the superclass constructor will be invoked. See the [superclass tests](https://github.com/munificent/wren/blob/master/test/language/constructor/superclass.wren) for more examples on calling the super constructor.

**Comment by ComFreek** [comment]
Wow - I feel dumb for not having tested this :smile: Thanks!

Maybe the docs on *Classes* should be updated to reflect the `super.method() / super()` functionality in general. Can I edit the HTML documentation files by hand or are they generated by some tool (although #85 suggests the opposite)?

**Comment by txgruppi** [comment]
I found some 'strange' things about `super`

1. `A does not implement 'new()'.`
   Maybe implementing the `new()` method in the `Object` class would solve this.
   I believe this will be useful for when I can't see the superclass, for
   example a superclass defined by the language which is controlling the vm.

```wren
class A {
}

class B is A {
  new() {
    super()
  }
}

new B()
```

2. `Command terminated`
   I know this code is wrong, I should not call `super()` on `A` but the thing
   is the message. It says `Command terminated` where it should say
   `Object does not implement 'new()'.` right? Anyway, 'fixing' the previous
   item will at least help to fix this.

```wren
class A {
  new() {
    super()
  }
}

class B is A {
}

class C is B {
  new() {
    super()
  }
}

new C()
```

Cheers

**Comment by bjorn** [comment]
@txgruppi I don't think it's strange at all for `super` to be an error if the superclass does not implement the current method. The thing with `new()` is that Wren uses `new` for the constructor without parameters, which is a different signature in Wren.

Your second issue seems to be a bug in Wren. Instead of saying `Object does not implement 'new()'`, the `A.new()` function ends up calling itself until the stack overflows.

**Comment by txgruppi** [comment]
@bjorn Sorry about the first one. I messed up with the `new` and `new()`.

**Comment by munificent** [comment]
> Maybe the docs on Classes should be updated to reflect the super.method() / super() functionality in general.

Yes, that would be great. I really need to finish off those docs.

>  The thing with `new()` is that Wren uses `new` for the constructor without parameters, which is a different signature in Wren.

The support for empty parameter lists (`()`) is pretty new. One of the corners that I noticed but didn't put a lot of thought into is what that means for the default constructor. Right now, `Object` still has `new` from the days before `()`. Maybe it should have `new()`. Or both?

> Your second issue seems to be a bug in Wren. Instead of saying Object does not implement 'new()', the A.new() function ends up calling itself until the stack overflows.

I think that's right. That looks like a bug.

**Comment by ComFreek** [comment]
@munificent When does it make sense to declare a method `fun()`? Or to put it differently, what is the difference between `fun()` and `fun`? I know they have different signatures and are therefore not interchangeable, but how do they differ on a semantic level?

If they don't differ, I'd always opt for the version without a parameter list. Hence, `Object` should only have `new` and nothing else.

**Comment by bjorn** [comment]
@ComFreek It seems the current guidance is that `fun` is used for non-modifying accessor methods, while `fun()` is used for modifying functions. Of course, `new` is generally not const nor it is an accessor, but I'd personally still prefer constructors without arguments to not require `()` (which I'm also used to from C++).

**Comment by munificent** [comment]
Yeah, I agree with @bjorn. My rule of thumb is that I leave off the `()` unless the method modifies the receiver in some visible way. So, `count`, `abs`, and `toString` all don't have `()`, but `clear()` on List does.

**Comment by MarcoLizza** [comment]
I remember writing on the subject earlier this year and we ended in agreeing with the convention that @munificent described. Sounded quite reasonable and clear...

**Comment by munificent** [comment]
I spent some time looking into this bug. It turns out super calls in the VM are fundamentally broken. Oops!

They don't do the right thing at all when you use `super` in an inherited method. I have some ideas on how to fix it, so that's good at least.

**Comment by munificent** [comment]
Right. The problem is that Wren currently dispatches `super` calls by looking up the method on the superclass of the *receiver*. In cases where the method is inherited, that's wrong. It should be the superclass of *the class where the method was defined*.

In your example, the execution looks like:

```
(new C).method() // Looks up receiver's class: C.
super.method() in C's method // Look's up receiver's superclass: B.
super.method() in B's method // Looks up receiver's superclass: B. WRONG!
```

On the last line, it should be looking up the superclass of the class where `method()` is defined. In this case, we're talking about the method in `B`, so it should look up `A`.

In other words, superclass calls are always *statically dispatched to a fixed class*, and not at all based on the receiver's class.
